tests/multiprocess_gpu_test.py

```diff
@@ -38,8 +46,80 @@
 except ImportError:
   pytest = None
 
+config.update("experimental_xmap_spmd_lowering", True)
 config.parse_flags_with_absl()
 
+X_SIZE = 4
+Y_SIZE = 2
+Z_SIZE = 2
+ALL_AXES = ("x", "y", "z")
+
+def jax_multiprocess_init():
+  # Hookup the ENV vars expected to be set already in the SLURM environment
+  coordinator_address = os.environ.get("SLURM_STEP_NODELIST", None)
+  if coordinator_address is not None and '[' in coordinator_address:
+    coordinator_address = coordinator_address.split('[')[0] + \
+                          coordinator_address.split('[')[1].split(',')[0]
+  num_tasks = os.environ.get("SLURM_NPROCS", None)
+  taskid = os.environ.get("SLURM_PROCID", None)
+  localid = os.environ.get("SLURM_LOCALID", None)
+
+  # fixing port since it needs to be the same for all the processes
+  port = "54321"
+
+  print(f"coord addr:port : {coordinator_address}:{port}\nTotal tasks: "
+        f"{num_tasks}\ntask id: {taskid}\nlocal id: {localid}")
+
+  assert (coordinator_address is None or num_tasks is None or
+          taskid is None) is False
+
+  jax.config.update("jax_cuda_visible_devices", localid)
+
+  jax.distributed.initialize(
+      coordinator_address=f'{coordinator_address}:{port}',
+      num_processes=int(num_tasks),
+      process_id=int(taskid))
+
+  print(f"Total devices: {jax.device_count()}, Total tasks: {int(num_tasks)}, "
+        f"Devices per task: {jax.local_device_count()}")
+  assert (jax.device_count() == int(num_tasks) * jax.local_device_count())
+
+def sorted_devices():
+  devices = sorted(jax.devices(), key=lambda d: (d.id, d.host_id))
+  if len(devices) != 16:
+    raise unittest.SkipTest("Test assumes that it runs on 16 devices (2 nodes)")
+  return devices
+
+@contextlib.contextmanager
+def use_default_mesh():
+  jax_multiprocess_init()
```

Comment:
Why do you need to init this everytime?

Can you just init it once in `def setUp(self)`?

Reply:
What if one of the test crashes and it messes up the multihost setup? Will the other tests continue to run? 

Right now, I'm running each test in its own subprocess using `--forked` flag in pytest which basically let's each test run unaffected if one of the other tests fails.

Reply:
Ohh I see. ok sgtm.

Reply:
cool